Reimplement net/cert error pages in browser-modern theme. Fixes #1072…

jsantell opened the pull request:
…. r=vporof

The location bar isn't updating on an error, even though Search.jsx is receiving new props, and componentShouldUpdate returning true upon this. Any ideas? Can handle in a follow up.

jsantell commented on .eslintrc:
```diff
@@ -72,7 +72,7 @@
 
     "react/display-name": ["error", { ignoreTranspilerName: true }],
-
+    "react/prefer-stateless-function": ["off"],
```
https://medium.com/modus-create-front-end-development/component-rendering-performance-in-react-df859b474adc#.u5375qyp7

victorporof replied:
No need for this if you add shouldComponentUpdate everywhere.

jsantell replied:
I don't understand -- my reasoning for (re)removing this is turning stateless function components into React components is tedious when needed, and there's no perf overhead for doing so

victorporof replied:
Stateless function components don't have `shouldComponentUpdate` methods by definition. We shouldn't use those unless they don't receive any props or do receive props that *aways* change.

victorporof replied:
See the Key Takeaways (TL;DR) in the article you linked to. We should always use pure components.

jsantell replied:
I think we're in agreement then in disabling the preference of stateless function components

victorporof replied:
Disabling it would stop telling you to turn this
```js
class Foo extends Component {
  render() { return <div/> }
}
```
into a stateless function component. That means we can't easily catch class components that don't have a `shouldComponentUpdate` just by linting. It's definitely not the way this rule was intended to be used though, so I guess I agree with removing it. I'd just like to have something that warns us when we have a class component that doesn't have a `shouldComponentUpdate`, no matter what warning that is.

jsantell commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -73,33 +80,75 @@ class Page extends Component {
+      }));
+    });
+
+    this.webview.addEventListener('did-stop-loading', () => {
```
Previously, this was marking the page as LOADED, even in event of failure.

victorporof commented on app/ui/browser-modern/views/browser/page/cert-error.jsx:
```diff
@@ -0,0 +1,90 @@
+});
+
+class CertErrorPage extends Component {
+  render() {
```
Need a shouldComponentUpdate here.

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,58 @@
+});
+
+class ErrorPage extends Component {
+  render() {
```
Add shouldComponentUpdate.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -60,6 +63,10 @@ class Page extends Component {
+        code: null,
+        description: null,
+        url: null,
+        certificate: null,
```
Why reset all of these to null on a new-window event?

victorporof replied:
We're already resetting the entire page on navigation, see the page sagas.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -73,33 +80,75 @@ class Page extends Component {
     });
 
+    this.webview.addEventListener('did-finish-loading', () => {
+      // Event fired only when page is successfully loaded
```
Nit: fullstop at the end of sentences.
Can this be emitted multiple times for every single frame?

victorporof replied:
If this can be emitted multiple times, then a LOADED state can be set *after* a FAILED state being set.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -73,33 +80,75 @@ class Page extends Component {
+        canGoBack: this.webview.canGoBack(),
+        canGoForward: this.webview.canGoForward(),
+        location: url,
+        title,
```
Why set the canGoBack and canGoForward props here?
Should the location be update here? The `did-navigate-in-frame` event is better for in-frame navigations. Anything else that isn't handled?

victorporof replied:
Should the location be update here? The `did-navigate-in-frame` event is better for in-frame navigations. Anything else that isn't handled?

victorporof replied:
Can the title ever be different here? Why update it when we have the `page-title-set` event?

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -143,17 +195,19 @@ Page.displayName = 'Page';
   pageIndex: PropTypes.number.isRequired,
-  onMount: PropTypes.func.isRequired,
+  pageLocation: PropTypes.string.isRequired,
+  pageState: PropTypes.instanceOf(PageState).isRequired,
```
Import from shared props for this.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -130,6 +179,9 @@ class Page extends Component {
         className={`browser-page ${PAGE_STYLE}`}>
+        <ErrorPage
+          hidden={this.props.pageState.load !== PageState.STATES.FAILED}
+          {...this.props.pageState.toJS()} />
```
Ugh, please let's not uselessly convert immutable js objects to js objects for no reason. Update the ErrorPage to handle immutable objects instead.

victorporof commented on app/ui/browser-modern/views/browser/page/net-error.jsx:
```diff
@@ -0,0 +1,54 @@
+});
+
+class NetErrorPage extends Component {
+  render() {
```
Needs a shouldComponentUpdate.

victorporof commented on app/ui/browser-modern/views/browser/page/net-error.jsx:
```diff
@@ -0,0 +1,54 @@
+import Style from '../../../../shared/style';
+
+const CONTAINER_STYLE = Style.registerStyle({
+  width: '700px',
```
What if the browser is not wide enough for this? This should either flex or use precentage or viewport dimensions.

victorporof commented on app/ui/browser-modern/views/browser/page/net-error.jsx:
```diff
@@ -0,0 +1,54 @@
+    return (
+      <div className={CONTAINER_STYLE}>
+        <h1>Could not connect</h1>
+        <p>Tofino can't find the server at <strong>{this.props.url}</strong></p>
```
Please don't use tags for styling :)

jsantell replied:
This was just an error page thrown together in the previous front end, we don't have any mockups so this was just made so we don't have a blank page.

victorporof commented on app/ui/browser-modern/views/browser/page/net-error.jsx:
```diff
@@ -0,0 +1,54 @@
+        <ul>
+          <li>
+            Check the address for typing errors such as <strong>ww</strong>.example.com
+            instead of <strong>www</strong>.example.com
```
Ditto here. Have a class instead of `strong`.

victorporof commented on app/ui/browser-modern/views/browser/page/cert-error.jsx:
```diff
@@ -0,0 +1,90 @@
+  url: PropTypes.string,
+  code: PropTypes.number,
+  description: PropTypes.string,
+  certificate: PropTypes.object,
```
Don't use object prop types. Define a shape in shared props, or use immutable.js objects all the way.

victorporof replied:
Aren't all these props required? Then mark them as such.

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,58 @@
+  height: '100%',
+  position: 'absolute',
+  color: 'rgb(66, 78, 90)',
+  backgroundColor: '#FBFBFB',
```
Use the css variables in the theme file for colors.

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,58 @@
+  url: PropTypes.string,
+  code: PropTypes.number,
+  description: PropTypes.string,
+  certificate: PropTypes.object,
```
Ditto, no object prop types please.

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,58 @@
+  code: PropTypes.number,
+  description: PropTypes.string,
+  certificate: PropTypes.object,
+  hidden: PropTypes.bool,
```
Ditto, required props should be marked.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -73,33 +80,75 @@ class Page extends Component {
+        const { error, certificate } = await getCertificateError(validatedURL);
+        state.certificate = certificate;
+        state.description = error;
+      }
```
All of this should be handled in a saga instead. Logic in component views should be kept at a minimum.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -73,33 +80,75 @@ class Page extends Component {
+      // error description can be retrieved from the main process. Just check
+      // here roughly if the error looks like a cert error.
+      if (Math.abs(errorCode) === 501 || /CERT/.test(errorDescription)) {
+        const { error, certificate } = await getCertificateError(validatedURL);
```
This should *definitely* be in a saga, it's causes side effect.

victorporof commented:
@jsantell Setting the location bar text is handled via a side-effect now. React-managed components prove to be difficult to handle especially in our case where there is complex interaction with the underlying dom element (<input>). As such, any interaction with this dom element is done via the corresponding side effect actions.

victorporof commented:
Clearing review, there are a few things that need to be addressed.

victorporof commented on app/ui/browser-modern/views/browser/page/cert-error.jsx:
```diff
@@ -0,0 +1,90 @@
+});
+
+const CERTIFICATE_STYLE = Style.registerStyle({
+  backgroundColor: '#ddd',
```
Background colors should be theme variables.

jsantell replied:
Alright, but I disagree with all these theme variables that are for some ugly, unmocked up, barely styled page as a placeholder for a blank page that seems like something broke

victorporof commented on app/ui/browser-modern/views/browser/page/cert-error.jsx:
```diff
@@ -0,0 +1,90 @@
+      <div className={CONTAINER_STYLE}>
+        <h1>Your connection is not secure</h1>
+        <p>
+          The owner of <strong>{this.props.url}</strong> has configured
```
No `<strong>`.

jsantell replied:
I just removed all <strong> tags

victorporof commented on app/ui/browser-modern/model/page-state.js:
```diff
@@ -14,6 +14,10 @@ import Immutable from 'immutable';
+  code: undefined,
+  description: undefined,
+  certificate: undefined,
+  url: undefined,
```
Why is there an `url` property in the page state? There's a `location` in the page model. Having both is confusing.

victorporof commented:
Sorry about all the comments 😨

jsantell commented:
@victorporof  Addressed most comments -- except the moving the certificate fetching to a saga, and calling the URL bar update action. I'd like to discuss how this should be done first, as I don't completely grok sagas yet and don't want to rewrite this again

victorporof commented:
@jsantell sure thing, let's talk about it when you get the time.

jsantell commented:
Added sagas for certificate fetching, and updating the URL bar via effect

victorporof commented on app/ui/browser-modern/model/shared-prop-types.js:
```diff
@@ -18,18 +18,29 @@ import ProfileModel from './profile';
+  validStart: PropTypes.string,
+  validExpiry: PropTypes.string,
+  fingerprint: PropTypes.string,
+});
```
Any reason this couldn't be a record?

jsantell replied:
Unlikely

victorporof commented on app/ui/browser-modern/sagas/page-sagas.js:
```diff
@@ -137,3 +141,8 @@ function* capturePage({ pageId, webview }) {
 }
+
+function* getCertificateError({ pageId, url }) {
+  const { error, certificate } = yield Certificate.getCertificateError(url);
```
This should be `yield call(Certificate.getCertificateError, url)`. I'm surprised this works as is.

victorporof commented on app/ui/browser-modern/views/browser/page/cert-error.jsx:
```diff
@@ -0,0 +1,95 @@
+});
+
+const CERTIFICATE_STYLE = Style.registerStyle({
+  backgroundColor: '--theme-error-page-cert-background-color',
```
This should be `var(--varname)`. Surprised this works otherwise.

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,62 @@
+  height: '100%',
+  position: 'absolute',
+  color: '--theme-error-page-color',
+  backgroundColor: '--theme-error-page-background-color',
```
Ditto here, `var(--varname)`

victorporof commented on app/ui/browser-modern/views/browser/page/error-page.jsx:
```diff
@@ -0,0 +1,62 @@
+    const code = Math.abs(this.props.pageState.get('code'));
+    const isCertError = code === 501 || // ERR_CERT_AUTHORITY_INVALID
+                        code === 129 || // ERR_SSL_WEAK_SERVER_EPHEMERAL_DH_KEY
+                        /CERT/.test(this.props.pageState.get('description')); // Catch all
```
So no reason to move this code into some utility module? It would make things easier to test.

jsantell replied:
We could, but spending any more days on reimplenting this feature that was previously landed is something that might drive me mad, and just loading helpful error pages is a substantial UX gain and this blocks other patches

victorporof replied:
:fearful:

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -97,26 +99,66 @@ class Page extends Component {
+      // here roughly if the error looks like a cert error.
+      if (Math.abs(errorCode) === 501 || /CERT/.test(errorDescription)) {
+        this.props.dispatch(PageEffects.getCertificateError(this.props.pageId, validatedURL));
+      }
```
There's lots and lots of difficult to test logic in here. I'd move most of it into utility functions in some module.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -97,26 +99,66 @@ class Page extends Component {
+      // May be emitted multiple times for every single frame.
+      // Only emitted when a failure occurs, unlike `did-stop-loading`.
+      // Is fired before did-finish-load and did-stop-loading when failure
+      // occurs.
```
Super nit: keep the "occurs" on the same line.

victorporof commented on app/ui/browser-modern/views/browser/page/index.jsx:
```diff
@@ -177,8 +223,20 @@ Page.displayName = 'Page';
-  pageId: PropTypes.string.isRequired,
   onMount: PropTypes.func.isRequired,
+  pageId: PropTypes.string.isRequired,
+  pageIndex: PropTypes.number.isRequired,
```
`pageIndex` isn't being used anywhere, remove it.

jsantell commented:
Fixed nits, made cert an Immutable record, any utility stuff should handle in a follow up, this patch is killing me
